core/drivers/dra7_rng.c

```diff
@@ -0,0 +1,217 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <keep.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/mutex.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#define	RNG_OUTPUT_L            0x0000
+#define	RNG_OUTPUT_H            0x0004
+#define	RNG_STATUS              0x0008
+#  define RNG_READY             BIT(0)
+#  define SHUTDOWN_OFLO         BIT(1)
+#define	RNG_INTMASK             0x000C
+#define	RNG_INTACK              0x0010
+#define	RNG_CONTROL             0x0014
+#  define ENABLE_TRNG           BIT(10)
+#define	RNG_CONFIG              0x0018
+#define	RNG_ALARMCNT            0x001C
+#define	RNG_FROENABLE           0x0020
+#define	RNG_FRODETUNE           0x0024
+#define	RNG_ALARMMASK           0x0028
+#define	RNG_ALARMSTOP           0x002C
+#define	RNG_LFSR_L              0x0030
+#define	RNG_LFSR_M              0x0034
+#define	RNG_LFSR_H              0x0038
+#define	RNG_COUNT               0x003C
+#define	RNG_OPTIONS             0x0078
+#define	RNG_EIP_REV             0x007C
+#define	RNG_MMR_STATUS_EN       0x1FD8
+#define	RNG_REV                 0x1FE0
+#define	RNG_SYS_CONFIG_REG      0x1FE4
+#  define RNG_AUTOIDLE          BIT(0)
+#define	RNG_MMR_STATUS_SET      0x1FEC
+#define	RNG_SOFT_RESET_REG      0x1FF0
+#  define RNG_SOFT_RESET        BIT(0)
+#define	RNG_IRQ_EOI_REG         0x1FF4
+#define	RNG_IRQSTATUS           0x1FF8
+
+#define RNG_CONTROL_STARTUP_CYCLES_SHIFT        16
+#define RNG_CONTROL_STARTUP_CYCLES_MASK         (0xffff << 16)
+
+#define RNG_CONFIG_MAX_REFIL_CYCLES_SHIFT       16
+#define RNG_CONFIG_MAX_REFIL_CYCLES_MASK        (0xffff << 16)
+#define RNG_CONFIG_MIN_REFIL_CYCLES_SHIFT       0
+#define RNG_CONFIG_MIN_REFIL_CYCLES_MASK        (0xff << 0)
+
+#define RNG_ALARMCNT_ALARM_TH_SHIFT             0x0
+#define RNG_ALARMCNT_ALARM_TH_MASK              (0xff << 0)
+#define RNG_ALARMCNT_SHUTDOWN_TH_SHIFT          16
+#define RNG_ALARMCNT_SHUTDOWN_TH_MASK           (0x1f << 16)
+
+#define RNG_CONTROL_STARTUP_CYCLES              0xff
+#define RNG_CONFIG_MIN_REFIL_CYCLES             0x21
+#define RNG_CONFIG_MAX_REFIL_CYCLES             0x22
+#define RNG_ALARM_THRESHOLD                     0xff
+#define RNG_SHUTDOWN_THRESHOLD                  0x4
+
+#define RNG_FRO_MASK    0xffffff
+
+#define RNG_REG_SIZE    0x2000
+
+register_phys_mem(MEM_AREA_IO_SEC, RNG_BASE, RNG_REG_SIZE);
+
+static struct mutex rng_mutex = MUTEX_INITIALIZER;
+
+static vaddr_t rng_base(void)
+{
+	static void *va __early_bss;
+
+	if (cpu_mmu_enabled()) {
+		if (!va)
+			va = phys_to_virt(RNG_BASE, MEM_AREA_IO_SEC);
+		return (vaddr_t)va;
+	}
+
+	return RNG_BASE;
+}
+
+uint8_t hw_get_random_byte(void)
+{
+	static vaddr_t rng;
+	static int pos;
+	static union {
+		uint64_t val;
+		uint8_t byte[8];
+	} random;
+	uint8_t ret;
+
+	mutex_lock(&rng_mutex);
+
+	if (!rng)
+		rng = rng_base();
```

Comment:
It's not exact bench-marking but the added delay is obvious when testing the probabilistic "Primality Algorithms" as part of the xtest suite. I can perform more detail analysis of this later.

I'll also switch to using a spinlock, even the longest path (fixing FRO failure path) is still only a few register reads so this should be safe (although an un-bounded 'while' loop in a spin-lock still concerns me slightly).